### ISSUE TITLE
Blank screen on KDE with VirtualGL on EGL with NVIDIA

### ISSUE THREAD

**Issue by ehfd** [VirtualGL/virtualgl#224]
Environment: Ubuntu 22.04 with KDE, using `vglrun +wm startplasma-x11` with Xvfb with `export VGL_DISPLAY="egl"`.

What I am supposed to see (In the exact same condition except when using Xorg with NVIDIA 515 drivers):
![image](https://user-images.githubusercontent.com/8457324/212337839-446961ee-11fe-4967-815c-aa4c1598b537.png)

What I see:
![image](https://user-images.githubusercontent.com/8457324/212336671-2e068738-7947-4463-95ad-378d4763375b.png)

Is there any other considerations for using QT desktop environments with VirtualGL?
I believe these issues may be affected by QT XCB.

**Comment by dcommander** [comment]
Hard to say.  Regular Qt applications work with VirtualGL, so I would need a better understanding of what Plasma is doing differently.  Unfortunately, the odds that I will be able to look into this anytime soon are close to zero.  TurboVNC doesn't officially support Plasma, so it falls into the category of "community support."  That means that I am happy to integrate patches that fix issues with it, but diagnosing those issues isn't something I can do outside of the context of a paid support contract.

**Comment by ehfd** [comment]
I needed the compositor off anyways, so it will be kept that way for now...

**Comment by dcommander** [comment]
Contact me through e-mail:
https://virtualgl.org/About/Contact

**Comment by ehfd** [comment]
There is one way to use VirtualGL + KWin Compositing: Set Rendering `Backend: XRender`, both OpenGL 3.0 and 2.1 fails now.

**Comment by ehfd** [comment]
XRender is not in Ubuntu 22.04.

Possible fix:

`export QT_XCB_GL_INTEGRATION=xcb_egl`
`export KWIN_OPENGL_INTERFACE=egl`

From https://linuxreviews.org/HOWTO_enable_10-bit_color_on_Linux#KDE_Plasma_5.x.x_And_The_Deepin_Desktop_Envionment_Won.27t_Start

I will tell you if it is successful, and if you should add that to the docs or not.

**Comment by ehfd** [comment]
```
export QT_XCB_GL_INTEGRATION=xcb_egl
export KWIN_OPENGL_INTERFACE=egl
```

I confirm that these are the requirements. @dcommander Please close at your convenience. But please do add this to the docs if possible.

**Comment by ehfd** [comment]
VirtualGL 3.0.90 does not fix the issue.

```
export QT_XCB_GL_INTEGRATION=xcb_egl
export KWIN_OPENGL_INTERFACE=egl
```

@dcommander Could you add these to the docs for future reference on Plasma 5? You can close this issue at your convenience.

**Comment by ehfd** [comment]
It's still broken. The last time I made it work was through llvmpipe. I guess it's something with the pixmaps that must be supported.

**Comment by dcommander** [comment]
Is there a way to tell Qt to use GLX (`QT_OPENGL=desktop`)?  I'm just wondering if this is another facet of #226, which also seems to be related to Qt trying and failing to use our new EGL/X11 front end.

**Comment by ehfd** [comment]
> I'm just wondering if this is another facet of https://github.com/VirtualGL/virtualgl/issues/226, which also seems to be related to Qt trying and failing to use our new EGL/X11 front end.

Possible, however the VLC issue arise additionally in 3.0.91. This issue is in 3.0.2 as well as 3.0.9x.

I am trying to make some time for reproducing this.

**Comment by ehfd** [comment]
Solved together with #226.

**Comment by ehfd** [comment]
![image](https://user-images.githubusercontent.com/8457324/226162406-d0cdb72a-9e45-4e5a-a7cc-3b08d204cddb.png)

**Comment by ehfd** [comment]
QT_OPENGL=desktop
changes nothing.

**Comment by dcommander** [comment]
Yeah, I personally observed the issue after the VLC issue was fixed, and that issue affected the EGL/X11 front end, whereas this issue affects the GLX front end with the EGL back end. I suspect that it's because the EGL back end doesn't yet implement `GLX_EXT_texture_from_pixmap`.

**Comment by ehfd** [comment]
What would be the steps needed for support?

**Comment by dcommander** [comment]
Contact me through e-mail:
https://virtualgl.org/About/Contact
We would first need to determine whether the addition of `GLX_EXT_texture_from_pixmap` would really solve the problem, then I can give you a formal estimate for the cost of developing it.

**Comment by dcommander** [comment]
Any news on this front?  It might be worth trying the latest pre-release build, which includes a fix for the EGL back end that affected other applications.

**Comment by ehfd** [comment]
Hi, Darrell. Always a pleasure to converse with you.
I'll try it right away. So, this commit is in the latest dev build (in AWS S3), I assume?

**Comment by dcommander** [comment]
Lastest 3.1.x Stable build:

https://virtualgl.org/DeveloperInfo/PreReleases

**Comment by ehfd** [comment]
I came back to test it. Took a bit of time until I came back.
It doesn't work with GLX in the default settings, but it does work with `KWIN_OPENGL_INTERFACE` set to `egl`.

**Comment by ehfd** [comment]
![image](https://github.com/VirtualGL/virtualgl/assets/8457324/c6a19169-e1b6-4749-b53a-4aa0d548474d)

Unfortunately, basically the same issue. `export QT_XCB_GL_INTEGRATION=xcb_egl` and `export KWIN_OPENGL_INTERFACE=egl` shows the above.

![image](https://github.com/VirtualGL/virtualgl/assets/8457324/359b7cba-6652-4caf-acdf-715f79749b3f)

With `QT_OPENGL=desktop` above.

**Comment by ehfd** [comment]
![image](https://github.com/VirtualGL/virtualgl/assets/8457324/45853da9-7c38-4619-b73a-4b457cbfb108)

I reproduce this situation with Chrome on KDE as well.

**Comment by ehfd** [comment]
Related to: https://gitlab.gnome.org/GNOME/gtk/-/issues/4815

**Comment by dcommander** [comment]
As an experiment, I disabled `GLX_EXT_texture_from_pixmap` in VirtualGL's GLX back end.  Plasma still works fine with the GLX back end if that extension is disabled, which casts doubt on my assertion that the lack of `GLX_EXT_texture_from_pixmap` coverage is responsible for the incompatibility between Plasma and the EGL back end.  I didn't see anything obvious in the QtBase source code that would point to a lack of coverage in VGL's EGL back end vs. its GLX back end, but maybe I'm looking in the wrong place.

**Comment by dcommander** [comment]
It's a shot in the dark, but it might be worth re-testing with the latest code in main.  Some missing framebuffer object direct state access functions were implemented, although I have no idea if KDE uses any of them.

**Comment by ehfd** [comment]
Okay, we no longer use this specific setup, but we'll reference this. Thanks!

**Comment by dcommander** [comment]
I can confirm that the issue is not fixed.  Oh well.
